Remove unnecessary uses of #[no_move]

eefriedman opened the pull request:
The patch makes RootCollection a bit safer by making the StackRootTLS hold
it in place.

RootedVec was doing an extremely delicate dance and just hoping nobody
messed it up; switch to a Box to be safe.

CodeGenRust seemed to be using no_move for no particularly good reason.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8286)

<!-- Reviewable:end -->

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented:
`no_move` is a lint from `rust-tenacious`, according to the original commit: https://github.com/servo/servo/commit/63714ebc5f41015876f3e3416a82eb99beec5e6b

eefriedman commented:
Oh... huh.  I completely missed that; I expected all the lints to be in components/plugins/lints.

In that case, it might make sense to leave RootedVec as-is for the moment.  The other changes still make sense, I think?

Manishearth commented:
`no_move` is no longer necessary for Root (which was its original use, but it's been removed from there now), but IIRC it's very necessary for RootedVec. Not sure about RootedCollection.

I'm wary of changing RootedVec without a discussion with nox and Ms2ger, it's a very delicate API.

Removing the no_move from bindings makes sense though, not sure how that got there.

cc @nox

nox commented:
RootedVec definitely still needs no_move, cf. https://github.com/servo/servo/issues/8159#issuecomment-150920083 where I tried to move one and got a panic at runtime.

nox commented:
@Manishearth It's weird that tenacious didn't catch that specific piece of code, btw.

Manishearth commented:
@nox tenacious only handles some cases, the issue is that Rust optimizes most moves into noops, but not always (actual in-memory move and compile-time "moved variable" are different), so to keep the lint from being extremely noisy the cases where it's usually optimized aren't really handled.

nox commented:
@Manishearth Yeah I understand that, but is RVO ever done to return from plain old code blocks, as opposed to function blocks?

Manishearth commented:
It should be.

It's easy enough to check for though, just add a check_block to no_move that ensures the retval isn't a nomove type.

eefriedman commented:
I'm pretty sure the version of RootedVec without no_move in this PR is safe... it's just a bit slower because of the extra heap allocation.

Manishearth commented:
r? @nox

nox commented:
I would rather keep RootedVec no_move and not behind a Box.

eefriedman committed:
Clean up some code related to #[no_move].

The patch makes RootCollection a bit safer by making the StackRootTLS hold
it in place.

The use of no_move in CodeGenRust was leftover from when roots couldn't
be moved.

eefriedman commented:
@nox Updated.

nox commented:
You forgot to remove "Misc cleanups around RootedVec." from the commit message. Otherwise, LGTM.

For reference, codegen used to mark dictionary structures as not movable because they can contain roots, which used to not be movable either.

-S-awaiting-review +S-needs-code-changes 
<!-- Reviewable comment -K2XFk1hEY5CeK249cv_:-1063597856 -->

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8286)
<!-- Sent from Reviewable.io -->

eefriedman commented:
@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 1a50fce has been approved by `nox`

<!-- @bors-servo r=nox 1a50fce67c840123447c9949c9692f4bb5828e5d -->

bors-servo commented:
:hourglass: Testing commit 1a50fce with merge 45a1cb0...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/319)

eefriedman commented:
@bors-servo retry (weird one-time Android build failure)

bors-servo commented:
:hourglass: Testing commit 1a50fce with merge 92f9e58...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/321), [gonk](http://build.servo.org/builders/gonk/builds/327), [linux-dev](http://build.servo.org/builders/linux-dev/builds/324), [linux-rel](http://build.servo.org/builders/linux-rel/builds/352), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/336), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/334), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/351)
